chainer/functions/array/get_item.py

```diff
@@ -37,22 +37,35 @@ def check_type_forward(self, in_types):
         type_check.expect(in_types[0].ndim >= valid_slice)
 
     def forward(self, xs):
-        self.retain_inputs(())
         ary = xs[0]
         self._in_shape = ary.shape
         self._in_dtype = ary.dtype
         return utils.force_array(ary[self.slices]),
 
-    def backward(self, xs, gys):
-        xp = cuda.get_array_module(*gys)
-        gy = gys[0]
+    def backward(self, indexes, gy):
+        return GetItemGrad(
+            self.slices, self._in_shape, self._in_dtype).apply(gy)
```

Comment:
Input shape and dtype can be obtained by `self.inputs[0].{shape,dtype}`, so you do not have to save them in forward.